### PR TITLE
feat: lint duplicated product names in dependencies

### PR DESCRIPTION
**Pull request by rsarv3006**
Resolves https://github.com/tuist/tuist/issues/6118

### Short description 📝

Add a duplicated target product name linter. Also checks for destinations as well. Thanks @kwridan for pointing that out.

### How to test the changes locally 🧐

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/testing-strategy) for reference).

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistGenerator/Linter/ProjectLinter.swift [hunk 1]
```diff
@@ -60,13 +60,35 @@ class ProjectLinter: ProjectLinting {
             .reduce(into: [String: Int]()) { $0[$1] = ($0[$1] ?? 0) + 1 }
             .filter { $0.value > 1 }
             .keys
+
+        let duplicatedTargetProductNames = project.targets
+            .reduce(into: [String: Int]()) {
+                let destinations = $1.destinations.sorted(by: { a, b in
+                    a.rawValue < b.rawValue
+                }).map(\.rawValue).joined(separator: ",")
```
Comment (fortmarek) [inline Sources/TuistGenerator/Linter/ProjectLinter.swift:68]: I don't think we should check destinations. As per @kwridan's comment [here](https://github.com/tuist/tuist/issues/6118#issuecomment-2059658504), it's fine to have duplicate product names as long as they don't end up being linked to the same target. Sure, supporting multiple destinations was one of the use cases for this, but there might be different valid scenarios that we'd want to allow.

Ideally, we should check for all [rootTargets](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/GraphTraverser.swift#L33) whether any of their [transitive dependencies](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Graph/GraphTraverser.swift#L108) has a duplicate product name.

Comment (rsarv3006) [reply Sources/TuistGenerator/Linter/ProjectLinter.swift:68]: Thanks for the clarification! I will update and push when I have a chance.

**Review by pepicrft — APPROVED** [review]

**Comment by pepicrft** [comment]
@fortmarek / @kwridan I pushed a commit implementing your suggestion. Let me know what you think.

File: Sources/TuistGenerator/Linter/GraphLinter.swift [hunk 2]
```diff
@@ -172,6 +173,32 @@ public class GraphLinter: GraphLinting {
...
 
+    private func lintDuplicatedProductNamesInDependencies(graphTraverser: GraphTraversing) -> [LintingIssue] {
+        return graphTraverser.targets().flatMap { projectPath, projectTargets in
+            return projectTargets.flatMap { targetName, _ in
+                var seenProductNames: Set<String> = []
+                var duplicatedProductNames: Set<String> = []
+                for productName in graphTraverser
+                    .allTargetDependencies(path: projectPath, name: targetName)
+                    .map(\.target.productNameWithExtension)
+                {
```
Comment (fortmarek) [inline Sources/TuistGenerator/Linter/GraphLinter.swift:184]: More of an open question, but do we know if this causes any performance hit? Each target loops always through all of its dependencies which, I'd think, can start to be expensive in large graphs. I'd consider memoising the individual subgraphs as we do in other places.

**Review by fortmarek — APPROVED** [review]
Looks good, my only concern is around performance here 👍

File: Sources/TuistGenerator/Linter/GraphLinter.swift (hunk 2 above)
Comment (pepicrft) [reply Sources/TuistGenerator/Linter/GraphLinter.swift:184]: I've added internal memoizing to the `allTargetDependencies` function.

**Comment by danthorpe** [comment]
@pepicrft @fortmarek I've bumped into this lint error in my project on v4.26. However, I'm wondering if there is an edge case regarding duplicate product names.

My project has the following characteristics, I have an iOS app target with a product name `"MyApp"`, a watch target which also has a product name of `"MyApp"`, and a ui test target which depends on the iOS app target.

I get the warning
> The target 'MyApp-PerformanceTest' has dependencies with the following duplicate product names: MyApp.app

However, if I remove the dependency from the UI test target - then it doesn't get its target application build setting configured correctly. And if I remove the dependency on the watch target - it doesn't get compiled with the app.

Is there some documentation which shows how to configure a project with a watch app & UI test targets? (I couldn't find any)

**Comment by KillianSmith0** [comment]
> @pepicrft @fortmarek I've bumped into this lint error in my project on v4.26. However, I'm wondering if there is an edge case regarding duplicate product names.
> 
> My project has the following characteristics, I have an iOS app target with a product name `"MyApp"`, a watch target which also has a product name of `"MyApp"`, and a ui test target which depends on the iOS app target.
> 
> I get the warning
> 
> > The target 'MyApp-PerformanceTest' has dependencies with the following duplicate product names: MyApp.app
> 
> However, if I remove the dependency from the UI test target - then it doesn't get its target application build setting configured correctly. And if I remove the dependency on the watch target - it doesn't get compiled with the app.
> 
> Is there some documentation which shows how to configure a project with a watch app & UI test targets? (I couldn't find any)

+1 here very similar configuration with tests and watch target. Once I remove the watch target from the app, it successfully builds. 

For reference: I have this open issue https://github.com/tuist/tuist/issues/6715
